dask/dataframe/methods.py

```diff
@@ -259,14 +259,18 @@ def concat(dfs, axis=0, join='outer', uniform=False):
 
     # Handle categorical index separately
     dfs0_index = dfs[0].index
-    if (isinstance(dfs0_index, pd.CategoricalIndex) or
-            (isinstance(dfs0_index, pd.MultiIndex) and
-             any(isinstance(i, pd.CategoricalIndex) for i in dfs0_index.levels))):
+
+    has_categoricalindex = (
+        isinstance(dfs0_index, pd.CategoricalIndex) or
+        (isinstance(dfs0_index, pd.MultiIndex) and
+         any(isinstance(i, pd.CategoricalIndex) for i in dfs0_index.levels)))
+
+    if has_categoricalindex:
         dfs2 = [df.reset_index(drop=True) for df in dfs]
         ind = concat([df.index for df in dfs])
     else:
         dfs2 = dfs
-        ind = None
+        ind = concat([df.index for df in dfs2])
```

Comment:
@jcrist not sure if you have a chance to glance over this again, but changing this to set `ind` here, rather than None, passed the tests. The only other place we maybe set `ind` was in https://github.com/dask/dask/pull/2963/files#diff-ed1bb5cb465ea067d764e4168024f1a3L319 and it does the same thing (operating on `dfs2`), so I think we're OK. I didn't take time to fully comprehend what's going on here though :)

Reply:
Sorry for the delay here. The reason for this logic was to avoid concatenating the indices twice.

If it's a categorical index (or a multi-index with categoricals) we need to:
- Drop the index
- Concatenate all the frames
- Concatenate all the indices
- Re-add the index

If it's not a categorical index, we should just let pandas handle the index concatenation when it's concatenating the frames.

---

I just looked at this problem, and it only fails for empty frames. Found a bug in pandas:

```python
In [1]: import pandas as pd

In [2]: x = pd.Categorical(['a', 'b', 'c'])

In [3]: df = pd.DataFrame({'x': x}, index=['m', 'n', 'o'])

In [4]: empty_frame = df.iloc[:0].copy()

In [5]: new_data = pd.Categorical(['x', 'y', 'z'])

In [6]: empty_new_data = new_data[:0]

In [7]: type(df.index)
Out[7]: pandas.core.indexes.base.Index

In [8]: df['x'] = new_data  # assign new data

In [9]: type(df.index)  # index remains the same type
Out[9]: pandas.core.indexes.base.Index

In [10]: type(empty_frame.index)
Out[10]: pandas.core.indexes.base.Index

In [11]: empty_frame['x'] = empty_new_data  # assign new data

In [12]: type(empty_frame.index)  # index type changes!
Out[12]: pandas.core.indexes.range.RangeIndex
```

---

The following patch to dask fixes this problem (there might be a cleaner solution):

```
diff --git a/dask/dataframe/methods.py b/dask/dataframe/methods.py
index 9a0989a..89d0758 100644
--- a/dask/dataframe/methods.py
+++ b/dask/dataframe/methods.py
@@ -290,6 +290,7 @@ def concat(dfs, axis=0, join='outer', uniform=False):
             not_cat = cat_mask[~cat_mask].index
             out = pd.concat([df[df.columns.intersection(not_cat)]
                              for df in dfs3], join=join)
+            temp_ind = out.index
             for col in cat_mask.index.difference(not_cat):
                 # Find an example of categoricals in this column
                 for df in dfs3:
@@ -308,6 +309,9 @@ def concat(dfs, axis=0, join='outer', uniform=False):
                                                          sample.cat.ordered)
                         parts.append(data)
                 out[col] = union_categoricals(parts)
+            # Pandas resets index type on data assignment if frame is empty
+            if not len(temp_ind):
+                out.index = temp_ind
             out = out.reindex(columns=cat_mask.index)
         else:
             # pandas may raise a RuntimeWarning for comparing ints and strs
```

Reply:
To clarify, the problem is that for empty frames (but only empty frames), 

```
out[col] = union_categoricals(parts)
```

resets the index type.

Reply:
> I just looked at this problem, and it only fails for empty frames. Found a bug in pandas:

Found an issue for this, filed by you :) https://github.com/pandas-dev/pandas/issues/17101.